### PR TITLE
consensus/istanbul: implement snapshot marshal/unmarshal

### PR DESCRIPTION
**Pull request by markya0616**

### CODE REVIEW

**Review by tailingchen — APPROVED** [review]

File: consensus/istanbul/backend/snapshot.go [hunk 1]
```diff
@@ -272,3 +273,49 @@ func (s *Snapshot) validators() []common.Address {
...
+	if err := json.Unmarshal(b, &j); err != nil {
+		return err
+	}
+
+	s.Epoch = j.Epoch
+	s.Number = j.Number
+	s.Hash = j.Hash
+	s.Votes = j.Votes
+	s.Tally = j.Tally
+	s.ValSet = validator.NewSet(j.Validators, j.Policy)
```
Comment (yutelin) [inline consensus/istanbul/backend/snapshot.go]: missing policy?

File: consensus/istanbul/backend/snapshot_test.go [hunk 2]
```diff
@@ -400,3 +402,54 @@ func TestVoting(t *testing.T) {
...
+	}
+	if !reflect.DeepEqual(snap.Votes, snap.Votes) {
+		t.Errorf("votes mismatch: have %v, want %v", snap1.Votes, snap.Votes)
+	}
+	if !reflect.DeepEqual(snap.Tally, snap.Tally) {
+		t.Errorf("tally mismatch: have %v, want %v", snap1.Tally, snap.Tally)
+	}
+	if !reflect.DeepEqual(snap.ValSet, snap.ValSet) {
+		t.Errorf("validator set mismatch: have %v, want %v", snap1.ValSet, snap.ValSet)
+	}
```
Comment (yutelin) [inline consensus/istanbul/backend/snapshot_test.go]: missing policy checking

**Review by yutelin — APPROVED** [review]

File: consensus/istanbul/backend/snapshot_test.go (hunk 2 above)
Comment (markya0616) [reply consensus/istanbul/backend/snapshot_test.go]: We don't need to check policy here. Policy is already in `ValSet`.

File: consensus/istanbul/backend/snapshot.go (hunk 1 above)
Comment (markya0616) [reply consensus/istanbul/backend/snapshot.go]: no

**Pull request merged**